corehq/apps/accounting/utils/downgrade.py

```
@@ -101,6 +105,80 @@ def _get_communication_model_context(domain, oldest_unpaid_invoice):
+    get_eligible_customer_billing_accounts_fn = get_accounts_with_customer_invoices_due_soon
+
+    @staticmethod
+    def is_subscription_eligible_for_process(subscription):
+        return subscription.plan_version.plan.edition != SoftwarePlanEdition.PAUSED
+
+    @classmethod
+    def _check_and_perform_action(cls, communication_model, context,
+                                  oldest_unpaid_invoice, total, today, subscription):
+        if cls._should_send_invoice_reminder(communication_model, oldest_unpaid_invoice):
+            context = cls._update_email_context(context, oldest_unpaid_invoice, total, today, subscription)
+            cls._send_reminder_email(oldest_unpaid_invoice, communication_model, context)
+
+    @staticmethod
+    def _should_send_invoice_reminder(communication_model, invoice):
+        return not communication_model.objects.filter(
+            invoice=invoice,
+            communication_type=CommunicationType.INVOICE_REMINDER,
+        ).exists()
+
+    @staticmethod
+    def _send_reminder_email(invoice, communication_model, context):
+        if invoice.is_customer_invoice:
+            account_name = invoice.account.name
+            bcc = None
+        else:
+            account_name = invoice.get_domain()
+            bcc = [settings.GROWTH_EMAIL]
+
+        subject = _(
+            "Your CommCare Billing Statement for {} is due in {} days".format(
+                account_name, DAYS_BEFORE_DUE_TO_TRIGGER_REMINDER
+            ))
+
+        send_html_email_async.delay(
+            subject,
+            invoice.get_contact_emails(include_domain_admins=True, filter_out_dimagi=True),
+            render_to_string('accounting/email/invoice_reminder.html', context),
+            render_to_string('accounting/email/invoice_reminder.txt', context),
+            cc=[settings.ACCOUNTS_EMAIL],
+            bcc=bcc,
+            email_from=get_dimagi_from_email())
+        communication_model.objects.create(
```

do we want to do this as a callback function in the task? will that work, is it picklable? or do we create a separate accounting tasks that sends the email synchronously and then when it is done, create the communication model

---

might be a good idea to also go through the old code and update it where we do this incorrectly

---

There is a really simple solution here which is just to replace the use of `corehq.apps.hqwebapp.tasks.send_html_email_async` in this file with `corehq.util.log.send_HTML_email`, which is what the `_async` function itself calls. What would you think of that?

---

yes, i think that makes sense. how would we handle failures and retries?

---

If something fails while sending the email it should raise an exception, which since it's synchronous it will bubble up to the try/catch in `run_action`, to get logged as a failure to downgrade (or send reminder) for this project space or billing account:
https://github.com/dimagi/commcare-hq/blob/47dd3ab285e421c865cc17487dfc4935d0db4b79/corehq/apps/accounting/utils/unpaid_invoice.py#L49-L57

I can't tell how necessary a retry mechanism would be (I'm not clear if that's typically handling a failure of the task related to celery, or a failure of the email to send for other reasons). I do see that in other places we use `send_HTML_email` we don't add retries, which makes me think that it should be fine.

But if you'd be more comfortable seeing this as a separate accounting task that explicitly links the email sending (with retries) to creating the `CommunicationHistory`, I can do that instead.

---

just because we do it poorly elsewhere doesn't mean it's a good idea. I think let's use `send_HTML_email` for now to speed up dev, and split it into a task as a followup PR. I think we should at least escape errors and log them in order to make sure the entire task for all of the downgrade/reminder actions doesn't fail as a result of an email error. A separate task is nice so that the email goes to the proper celery queue for emails and it doesn't block this task as sometimes that can be a time consuming action

---

Okay, here's the ticket I created for follow-ups, including this task. https://dimagi.atlassian.net/browse/SAAS-16820